controller/volume_controller.go

```diff
@@ -2326,8 +2343,42 @@ func (vc *VolumeController) switchActiveReplicas(rs map[string]*longhorn.Replica
 	return nil
 }
 
-func (vc *VolumeController) isResponsibleFor(v *longhorn.Volume) bool {
-	return isControllerResponsibleFor(vc.controllerID, vc.ds, v.Name, v.Spec.NodeID, v.Status.OwnerID)
+func (vc *VolumeController) isResponsibleFor(v *longhorn.Volume, defaultEngineImage string) bool {
+	log := getLoggerForVolume(vc.logger, v)
+	readyNodesWithDefaultEI, err := vc.ds.ListReadyNodesWithEngineImage(defaultEngineImage)
+	if err != nil {
+		log.WithError(err).Warn("error while checking isResponsibleFor")
+		return false
+	}
+
+	isResponsible := isControllerResponsibleFor(vc.controllerID, vc.ds, v.Name, v.Spec.NodeID, v.Status.OwnerID)
+	if len(readyNodesWithDefaultEI) == 0 {
+		return isResponsible
+	}
+
+	if !isResponsible {
+		preferredOwnerEngineAvailable, _, err := vc.ds.CheckEngineImageReadiness(defaultEngineImage, v.Spec.NodeID)
+		if err != nil {
+			log.WithError(err).Warn("error while checking isResponsibleFor")
+			return false
+		}
+		currentOwnerEngineAvailable, _, err := vc.ds.CheckEngineImageReadiness(defaultEngineImage, v.Status.OwnerID)
+		if err != nil {
+			log.WithError(err).Warn("error while checking isResponsibleFor")
+			return false
+		}
+		if !preferredOwnerEngineAvailable && !currentOwnerEngineAvailable {
+			isResponsible = true
+		}
+	}
+
+	if isResponsible {
```

Comment:
Unless you are doing this because you are saying `len(readyNodesWithDefaultEI) == 0` we need to process the volume if there are no nodes available with an engine image. If so add a comment above that line, explaining that case.

Reply:
This can also be refactored to, flattened and a lot less conditionals :)
If you need to deal with the error cases, do them above, no need to if gate the calls to the ds.

```
currentOwnerEngineAvailable, err := 
preferredOwnerEngineAvailable, err :=
needToClaimOwnership := !preferredOwnerEngineAvailable && !currentOwnerEngineAvailable

currentNodeEngineAvailable, _, _ := vc.ds.CheckEngineImageReadiness(defaultEngineImage, vc.controllerID)
return (isResponsible && currentNodeEngineAvailable)  || needToClaimOwnership

Reply:
I have flattened the function

> consider moving this check after the call to isResponsible.
Then you can log an appropriate debug call and requeue the item.
example, node is responsible for processing volume %v but is missing engine image %v requeueing volume

I think we don't need to re-queue the item incase the current node is not ready. Just leave it to another ready node to pick up the object. How do you think?
Also, I think it is a bit cleaner if we keep all logic inside `isResponsibleFor` function so on the outside we can just `return nil` if the function `isResponsibleFor` returns `false`